### PR TITLE
fix(dot/sync): Pass logLvl to sync and digest

### PR DESCRIPTION
**Pull request by mittal-parth**
## Changes

Passes the logLevel also to the sync and digest service constructors.

This was required since the log level was not being set.

For instance, trace logs were not coming with sync. But now:

```bash
2024-12-12T23:29:31+05:30 INFO     receiving a block announce handshake from 12D3KooWBVTyMEvnQWhgjx6ZQ12SDXe8WEhZrkaaTLQEhLHVhBfs	pkg=sync
2024-12-12T23:29:31+05:30 TRACE    potential worker added, total in the pool 1	pkg=sync

```

This was required for:

- digest 
- sync 
- [x] runtime - couldn't find the appropriate handler for this where configs are set @haikoschol/maintainers need your input here.
- [x] wasmer - couldn't find the appropriate handler for this where configs are set

Other services already do it correctly:

- core
- network
- rpc
- babe
- grandpa
- state

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues

#4251

### CODE REVIEW

File: dot/sync/service.go
Comment (haikoschol) [inline dot/sync/service.go]: Was this added for testing?
```suggestion
```

**Review by haikoschol — CHANGES_REQUESTED** [review]
Thanks! LGTM, just this one extra log line to remove.

Regarding runtime/wasmer:
It looks like the `nodeBuilder` passes whatever is set under `wasmer` on the CLI as the log level for the runtime. This was probably forgotten to change when Gossamer switched from wasmer to wazero.

There is also a log level config for `runtime` though (which seems to be unused). IMHO it should be change so that this one is used by the `nodeBuilder` and `LogConfig.Wasmer` removed.

I'll leave it up to you to handle it in one of the following ways:
- Make the change in this PR
- Create an issue describing what needs to be done and linking it from this PR
- Do nothing about it (please let me know, I'll create the issue then)

File: dot/sync/service.go
Comment (mittal-parth) [reply dot/sync/service.go]: Right, my bad

**Comment by mittal-parth** [comment]
@haikoschol thanks!

I am a little confused here. This is what I understand.

- WasmInterpreter is a configuration.
- Wazero is a specific implementation of a WasmInterpreter.
- Runtime is the execution instance created by a WasmInterpreter (like Wazero) to run the blockchain logic.

Now since we have only one type of WasmInterpreter supported here, setting the log level for runtime and wazero would technically be the same?

OR

https://github.com/ChainSafe/gossamer/blob/7d3be4932f405e94bf1c689145ef24b2c10073a4/dot/services.go#L183-L198

should become

```diff
- wasmerLogLevel, err := log.ParseLevel(config.Log.Wasmer)
+ runtimeLogLvl, err := log.ParseLevel(config.Log.Runtime)
  if err != nil {
-     return nil, fmt.Errorf("failed to parse wasmer log level: %w", err)
+     return nil, fmt.Errorf("failed to parse runtime log level: %w", err)
  }
  switch config.Core.WasmInterpreter {
  case wazero_runtime.Name:
      rtCfg := wazero_runtime.Config{
          Storage:     ts,
          Keystore:    ks,
-         LogLvl:      wasmerLogLevel,
+         LogLvl:      runtimeLogLvl,
          NodeStorage: ns,
          Network:     net,
          Transaction: st.Transaction,
          Role:        config.Core.Role,
          CodeHash:    codeHash,
      }
```

The latter would mean that we abstract the logLvl from the user and keep it agnostic of the choice of WasmInterpreter - whatever one specifies for runtime, becomes the log level. This also makes sense, since at a time one only runs one kind of Interpeter?

**Comment by haikoschol** [comment]
> The latter would mean that we abstract the logLvl from the user and keep it agnostic of the choice of WasmInterpreter - whatever one specifies for runtime, becomes the log level. This also makes sense, since at a time one only runs one kind of Interpeter?

Yes, that's what I meant: To only have one attribute on the `--log` flag and to call it `runtime` instead of whatever implementation we use to run the wasm code.

**Comment by dimartiro** [comment]
Oh this is a great contribution! I discovered this issue a few days ago while debugging, but I didn’t have time to fix it. Thank you, @mittal-parth!
I’ll approve it right after you address @haikoschol’s comments, and please don’t forget to rebase with the development branch. 🙌

**Comment by mittal-parth** [comment]
@haikoschol @dimartiro pushed changes, rebased off development.

**Comment by haikoschol** [comment]
> @haikoschol @dimartiro pushed changes, rebased off development.

Thanks! Changes look good now. Also thanks for updating the documentation (in more places than should be necessary).

Just two small fixes required before we can merge:

1. You need to regenerate the `nodeBuilder` mock, since the API changed.

You can see the commands used to generate it in [`mocks_generate_test.go`](https://github.com/ChainSafe/gossamer/blob/fd5af5d0ba7d50bfea4b53ebcab45d7e6ab467bc/dot/mocks_generate_test.go#L8). This requires installing [`mockgen`](https://github.com/uber-go/mock). Looks like we're still using `v4` in CI, but the latest version should work as well.

```
$ go install go.uber.org/mock/mockgen@v0.4.0
# in the repo root:
$ mockgen -source=./dot/node.go -destination=./dot/mock_node_builder_test.go -package=dot
```

2. After regenerating the mock, run `make lint` locally so you don't have to wait for CI. There's a few issues regarding `gofmt` and line lengths.

**Comment by haikoschol** [comment]
> > @haikoschol @dimartiro pushed changes, rebased off development.
> 
> Thanks! Changes look good now. Also thanks for updating the documentation (in more places than should be necessary).
> 
> Just two small fixes required before we can merge:

nevermind 😄✅

**Review by haikoschol — APPROVED** [review]

**Comment by mittal-parth** [comment]
> You need to regenerate the nodeBuilder mock, since the API changed.

Ah thanks. I realised this later and manually made the required changes. 
When I now run this command locally, only diff is in the documentation at the top.

```diff
// Code generated by MockGen. DO NOT EDIT.
- // Source: node.go
+ // Source: ./dot/node.go
//
// Generated by this command:
- //	mockgen -source=node.go -destination=mock_node_builder_test.go -package=dot
+ //	mockgen -source=./dot/node.go -destination=./dot/mock_node_builder_test.go -package=dot
```

Do you want me to push this? 

> make lint

done

**Comment by haikoschol** [comment]
> > You need to regenerate the nodeBuilder mock, since the API changed.
> 
> Ah thanks. I realised this later and manually made the required changes. When I now run this command locally, only diff is in the documentation at the top.
> 
> ```diff
> // Code generated by MockGen. DO NOT EDIT.
> - // Source: node.go
> + // Source: ./dot/node.go
> //
> // Generated by this command:
> - //	mockgen -source=node.go -destination=mock_node_builder_test.go -package=dot
> + //	mockgen -source=./dot/node.go -destination=./dot/mock_node_builder_test.go -package=dot
> ```
> 
> Do you want me to push this?

No, let's keep it as is. Didn't think to check whether running the command as I suggested would change this comment. Better to keep it as is. I think the way those mocks were generated is a bit outdated compared to the newer ones. But it works so no need to change it.

**Review by axaysagathiya — APPROVED** [review]
Looks great! Thanks for your contribution! 🎉

**Pull request merged**
